File: DSA/src/datastructures/linkedlists/dual/DoubleLinkedList.java
package datastructures.linkedlists.dual;

import datastructures.nodes.dual.DoubleNode;

public class DoubleLinkedList<T> {
    private DoubleNode<T> head;
    private DoubleNode<T> tail;

    public DoubleLinkedList() {
        this.head = null;
        this.tail = null;
    }

    // O(1)
    public void shift(T data) {
        if (this.head == null) {
            init(data);
            return;
        }
        addToStart(data);
    }

    // O(1)
    public void append(T data) {
        if (this.head == null) {
            init(data);
            return;
        }
        addToEnd(data);
    }

    // O(1)
    public T unshift() {
        if (isEmpty()) return null;
        T toBeRemovedItem = this.head.getData();
        if (this.head == this.tail) {
            clear();
        } else {
            deleteFromStart();
        }
        return toBeRemovedItem;
    }

    // O(1)
    public T pop() {
        if (isEmpty()) return null;
        T toBeRemovedItem = this.tail.getData();
        if (this.head == this.tail) {
            clear();
        } else {
            deleteFromEnd();
        }
        return toBeRemovedItem;
    }

    // O(n)
    public void deleteData(T data) {
        if (isEmpty() || !doesExist(data)) return;
        if (this.head.getData().equals(data)) {
            deleteFromStart();
            return;
        }
        if (this.tail.getData().equals(data)) {
            deleteFromEnd();
            return;
        }
        DoubleNode<T> cursor = this.head;
        while (cursor != null) {
            if (cursor.getData().equals(data)) break;
            cursor = cursor.getNext();
        }
        cursor.getPrev().setNext(cursor.getNext());
        cursor.getNext().setPrev(cursor.getPrev());
        cursor = null;
    }

    // O(n)
    public String toString() {
        String output = "Next: [ Head -> ";

        DoubleNode<T> cursor = this.head;
        while (cursor != null) {
            output += cursor.getData().toString() + " -> ";
            cursor = cursor.getNext();
        }
        output += "Tail -> ]\n";
        output += "Prev: [ Tail -> ";
        cursor = this.tail;
        while (cursor != null) {
            output += cursor.getData() + " -> ";
            cursor = cursor.getPrev();
        }
        output += "Head -> ]";
        return output;
    }

    // O(1)
    public boolean isEmpty() {
        return this.head == null;
    }

    // O(1)
    public void clear() {
        this.head = null;
        this.tail = null;
    }

    // O(n)
    public boolean doesExist(T data) {
        if (isEmpty()) return false;
        DoubleNode<T> cursor = this.head;
        while (cursor != null) {
            if (cursor.getData().equals(data)) return true;
            cursor = cursor.getNext();
        }
        return false;
    }

    private void addToStart(T data) {
        DoubleNode<T> newNode = new DoubleNode<>(data);
        newNode.setNext(this.head);
        this.head.setPrev(newNode);
        this.head = newNode;
    }

    private void deleteFromStart() {
        this.head = this.head.getNext();
        this.head.setPrev(null);
    }

    private void addToEnd(T data) {
        DoubleNode<T> newNode = new DoubleNode<>(data);
        this.tail.setNext(newNode);
        newNode.setPrev(this.tail);
        this.tail = newNode;
    }

    private void deleteFromEnd() {
        this.tail = this.tail.getPrev();
        this.tail.setNext(null);
    }

    private void init(T data) {
        this.head = new DoubleNode<>(data);
        this.tail = head;
    }
}

File: DSA/DataStructures/LinkedLists/DoubleLinkedListNS/DoubleLinkedList.cs
using System.Collections;
using DSA.DataStructures.Nodes.DoubleNodeNS;

namespace DSA.DataStructures.LinkedLists.DoubleLinkedListNS
{
  public class DoubleLinkedList<T> : IDoubleLinkedList<T>, IEnumerable<T>
  {
    private DoubleNode<T>? head;
    private DoubleNode<T>? tail;
    private int length;

    public DoubleLinkedList()
    {
      this.head = null;
      this.tail = null;
      this.length = 0;
    }

    public void AddEnd(T item)
    {
      IncrementLength();
      if (this.head == null)
      {
        InitializeList(item);
        return;
      }
      DoubleNode<T> newNode = new(item);
      newNode.Prev = this.tail;
      this.tail!.Next = newNode;
      this.tail = newNode;
    }

    public void AddStart(T item)
    {
      IncrementLength();
      if (this.head == null)
      {
        InitializeList(item);
        return;
      }
      DoubleNode<T> newNode = new(item);
      newNode.Next = this.head;
      this.head.Prev = newNode;
      this.head = newNode;
    }

    public T DeleteEnd()
    {
      if (this.IsEmpty()) if (this.IsEmpty()) throw new Exception("Cannot remove from empty list");
      this.DecrementLength();
      T deletedItem = this.tail!.Data!;
      if (this.head == this.tail)
      {
        this.Clear();
      }
      else
      {
        DoubleNode<T>? deletedNode = this.tail;
        this.tail = this.tail.Prev;
        this.tail!.Next = null;
        deletedNode.Prev = null;
        deletedNode = null;
      }
      return deletedItem;
    }

    public bool DeleteItem(T item)
    {
      if (!this.DoesExist(item)) return false;
      if (this.head!.Data!.Equals(item))
      {
        this.DeleteStart();
        return true;
      }
      if (this.tail!.Data!.Equals(item))
      {
        this.DeleteEnd();
        return true;
      }
      DecrementLength();
      DoubleNode<T>? cursor = this.head;
      while (!cursor!.Data!.Equals(item))
      {
        cursor = cursor.Next;
      }
      cursor.Prev!.Next = cursor.Next;
      cursor.Next!.Prev = cursor.Prev;
      cursor = null;
      return true;
    }

    public T DeleteStart()
    {
      if (this.IsEmpty()) throw new Exception("Cannot remove from empty list");
      DecrementLength();
      T removedItem = this.head!.Data!;
      if (this.head == this.tail)
      {
        this.Clear();
      }
      else
      {
        DoubleNode<T>? firstNode = this.head;
        this.head = this.head.Next;
        firstNode.Next = null;
        firstNode = null;
        this.head!.Prev = null;
      }

      return removedItem;
    }

    public bool DoesExist(T item)
    {
      if (this.IsEmpty()) return false;
      DoubleNode<T> cursor = this.head!;

      while (cursor != null)
      {
        if (cursor.Data!.Equals(item)) return true;
        cursor = cursor.Next!;
      }
      return false;
    }

    public bool IsEmpty()
    {
      return this.Length() == 0;
    }

    public int Length()
    {
      return this.length;
    }

    public override string ToString()
    {
      string output = "[ Head -> ";
      DoubleNode<T> cursor = this.head!;
      while (cursor != null)
      {
        output += $"{cursor.Data} -> ";
        cursor = cursor.Next!;
      }
      output += "Tail -> ]\n";
      output += "[ Tail -> ";
      cursor = this.tail!;

      while (cursor != null)
      {
        output += $"{cursor.Data} -> ";
        cursor = cursor.Prev!;
      }

      output += "Head -> ]";

      return output;
    }

    public IEnumerator<T> GetEnumerator()
    {
      DoubleNode<T> cursor = this.head!;
      while (cursor != null)
      {
        yield return cursor.Data!;
        cursor = cursor.Next!;
      }
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
      return this.GetEnumerator();
    }

    private void IncrementLength()
    {
      this.length++;
    }
    private void DecrementLength()
    {
      this.length--;
    }
    private void InitializeList(T item)
    {
      this.head = new DoubleNode<T>(item);
      this.tail = this.head;
    }
    private void Clear()
    {
      this.head = null;
      this.tail = null;
    }
  }
}